corehq/apps/app_manager/models.py

```diff
@@ -343,9 +343,17 @@ def get_action_properties(self, action):
                 yield name, path
 
 
-class UpdateCaseAction(FormAction):
+class SmartCaseUpdate(DocumentSchema):
+    question_path = StringProperty()
+    update_mode = StringProperty(
+        choices=["always", "edit"],
+        default="always"
+    )
+
 
-    update = DictProperty()
+class UpdateCaseAction(FormAction):
+    # TODO: migrate from dict(property_name: question_path) to dict(property_name: SmartCaseUpdate)
+    update = SchemaDictProperty(SmartCaseUpdate)
```

Comment:
This may already be part of your plan, but to make this code backwards compatible while the migration is happening, you'll want to add a custom `wrap` method to `FormActions` that handles the plain old dict. [Here's](https://github.com/dimagi/commcare-hq/blob/d08c8e968de0c963a78f1b8d89281f6cc156f81c/corehq/apps/app_manager/models.py#L568-L580) an example of a custom wrap - `data` is the raw JSON doc that you can manipulate. So you could check if the values in `update` are plain strings and if so, transform them into `SmartCaseUpdate`s.

If you do that, I think you can merge this PR, then migrate existing apps at your leisure, and then just do a quick cleanup PR to remove the custom wrap method.

(And something similar for `AdvancedAction`)

Reply:
Wrap functions were just added, thanks for the idea and let me know what you think!